### ISSUE TITLE
Temporal RNN-RBM model

### ISSUE THREAD

**Issue by qwertzdenek** [Element-Research/rnn#168]
Hi,
I am working on temporal music RNN-RBM model that was described in http://arxiv.org/abs/1206.6392. First implementation problem is that the biases depends on time step Linear operation so I can't do rbm simple like this:

```
propup = nn.Sequential()
propup:add(nn.Linear(n_inputs, n_hidden))
propup:add(nn.Sigmoid())
```

I am not sure how to do some pragmatic solution. I started to hardcode recurrent RBM but RNN->RBM gradients looks tricky. One possible solution is to split whole network and pretrain it with some script. Is it really necessary?

Next problem is Contrastive Divergence. Am I right to do it on _updateGradInput_ and return _gradWeight_, _gradVisibleBias_ _gradHiddenBias_? Problem is these gradients are used in many places. How will BPTT work on W' and W'' edges? I know about rbm_toolbox but this code is hardly reusable for such task.

I am newbie with rnn so want to know whether I can use this package.

**Comment by nicholas-leonard** [comment]
@qwertzdenek RBMs aren't really my thing. But I will try to be of help. Consulting the paper, I agree that the bias will be complicated to implement. You will require some `nn` module kung-fu to build a model like that. One way to do it is to use `LinearNoBias + CAddTable`. The `CAddTable` will take two inputs : the output of `LinearNoBias` and the bias. The bias is a the result of another `LinearNoBias(hiddensize, outputsize)` applied to the hidden state. Building the network will be complicated (but fun). 

As for contrastive divergence, you might need to implement your own RBM module to get that working. Ask around on the torch forums to find out if there are any existing RBM implementations in torch.

You can use the rnn package for your kind of project, but you will need to build/find an RBM module and do some non-standard `nn` module stacking.

**Comment by qwertzdenek** [comment]
Thanks for tip. I published code at https://github.com/qwertzdenek/elevator-music-generator/blob/sequencer/train.lua#L240 . I had to split `rnn` and RBM part as `mlp`, because they have inputs from the different times. I have problem with _optim_ training, because It does not changing any of the trainable parameters. Can it be caused by the `Module.flatten()` of already flattened Tensors?

**Comment by nicholas-leonard** [comment]
@qwertzdenek Try this instead of https://github.com/qwertzdenek/elevator-music-generator/blob/sequencer/train.lua#L263-L267: 
```lua
params, gradParams = nn.Container():add(rnn):add(mlp):getParameters()
```

**Comment by qwertzdenek** [comment]
Thanks, it works!
